app/worker/worker.go

```diff
@@ -86,56 +91,137 @@ func InitAndRun(
 		_ = unsubscribe()
 	}()
 
-	var msg interface{}
-	sem := make(chan bool, concurrency)
 	quit := make(chan os.Signal, 1)
 	signal.Notify(quit, unix.SIGINT, unix.SIGTERM)
+
+	// Spawn worker pool
+	var wg = new(sync.WaitGroup)
+	wg.Add(concurrency)
+	for i := 0; i < concurrency; i++ {
+		workerID := i
+		go func() {
+			defer func() {
+				wg.Done()
+				wg.Wait()
+				// Make sure context is cancelled when all workers are down
+				cancel()
+			}()
+			wl := l.F(log.Ctx{"worker_id": workerID})
+			wCtx := log.WithContext(ctx, wl)
+
+			wl.Info("worker starting up")
+			workerMain(wCtx, channel, natsClient, dataStore)
+			wl.Info("worker shut down")
+		}()
+	}
+
+	// Run until a SIGTERM or SIGINT is received or the entire pool is down
+	select {
+	case <-quit:
+	case <-ctx.Done():
+	}
+	// Notify workers that we're done
+	close(channel)
+	// Wait for all workers to finish their current task
+	wg.Wait()
+	return nil
+}
+
+// reusableTimer allows using the time.After interface without reallocations.
+type reusableTimer time.Timer
+
+func (t *reusableTimer) After(d time.Duration) <-chan time.Time {
+	tt := (*time.Timer)(t)
+	if !tt.Stop() {
+		<-tt.C
+	}
+	tt.Reset(d)
+	return tt.C
+}
```

Comment:
For the sake of readability, this could go to an independent file (or package, maybe)?

Reply:
Can do.

Reply:
Done.